### PR TITLE
Fix S6673 FP: Argument name contains placeholder name

### LINKED ISSUE #8769: New rule S6673: Log message template placeholders should be in the right order
See https://github.com/SonarSource/rspec/pull/2563

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Fixes #8769

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.Test/TestCases/LoggingTemplatePlaceHoldersShouldBeInOrder.cs [hunk 1]
```diff
@@ -75,6 +75,9 @@ public void PropertyAccess(ILogger logger, Person person, List<Person> people)
         logger.LogInformation("Father: {Father}, Mother: {Mother}", person.Mother.LastName, person.Father.LastName);    // FN
         logger.LogInformation("People: {Count} {People}", people.Count, people.Select(x => x.FirstName + " " + x.LastName));
 
+        var someFirstName = person.FirstName;
+        logger.LogInformation("[{PersonFirstName}] Disposing endpoint {FirstName}", person.FirstName, someFirstName);   // Compliant
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/LoggingTemplatePlaceHoldersShouldBeInOrder.cs:79]: Could you please add one more test case with:
- fields:
```csharp
 _logger.LogInformation("[ServerConnector][{SystemAddress}] Disconnected from {Address}", _system.Address, 
_address); 
```
- properties
```csharp
 _logger.LogDebug("[{SystemAddress}] Disposing endpoint {Address}", System.Address, 
RemoteAddress);
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs [hunk 2]
```diff
@@ -97,7 +98,7 @@ placeholderValue switch
         {
             MemberAccessExpressionSyntax memberAccess => MatchesName(placeholderName, memberAccess.Name),
             CastExpressionSyntax cast => MatchesName(placeholderName, cast.Expression),
-            NameSyntax name => new string(placeholderName.Where(char.IsLetterOrDigit).ToArray()).Equals(name.GetName(), StringComparison.OrdinalIgnoreCase),
+            NameSyntax name => placeholderName.Equals(name.GetName(), StringComparison.OrdinalIgnoreCase) || (placeholderName.Length >= 5 && name.GetName().IndexOf(placeholderName, StringComparison.OrdinalIgnoreCase) >= 0),
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs:101]: I'm not sure I understand where 5 is coming from.

Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs]: Removed it.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/LoggingTemplatePlaceHoldersShouldBeInOrder.cs [hunk 3]
```diff
@@ -5,128 +5,159 @@
...
-                                                                                                                        // Secondary @-1
+        logger.LogInformation("Person: {FirstName} {Age}", person.FirstName, person.Age);       // Compliant
+        logger.LogInformation("Person: {Age} {FirstName}", person.FirstName, person.Age);       // Noncompliant
+                                                                                                // Secondary @-1
+
         logger.LogInformation("Father: {Father}, Mother: {Mother}", person.Father.LastName, person.Mother.LastName);    // Compliant
         logger.LogInformation("Father: {Father}, Mother: {Mother}", person.Mother.LastName, person.Father.LastName);    // FN
+
+        logger.LogInformation("Person: {FirstName} {LastName}", person.FirstName, person.LastName);     // Compliant
+        logger.LogInformation("Person: {FirstName} {LastName}", person.LastName, person.FirstName);     // FN
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/LoggingTemplatePlaceHoldersShouldBeInOrder.cs:101]: Maybe I've missed the cases but I couldn't find just partial words, like:
```csharp
logger.LogInformation("Person: {FirstName} {LastName}", first, last);
logger.LogInformation("Person: {FirstName} {LastName}", last, first);
logger.LogInformation("Person: {FirstName} {LastName}", last, name);
logger.LogInformation("Person: {FirstName} {LastName}", name, name);
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs [hunk 4]
```diff
@@ -75,15 +75,16 @@ private static IEnumerable<SyntaxNode> PlaceholderValues(InvocationExpressionSyn
 
     private static SyntaxNode OutOfOrderPlaceholderValue(MessageTemplatesParser.Placeholder placeholder, int placeholderIndex, ImmutableArray<SyntaxNode> placeholderValues)
     {
-        if (placeholderIndex < placeholderValues.Length && MatchesName(placeholder.Name, placeholderValues[placeholderIndex]) is not false)
+        var filteredName = new string(placeholder.Name.Where(char.IsLetterOrDigit).ToArray());
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs]: Removed it.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/LoggingTemplatePlaceHoldersShouldBeInOrder.cs [hunk 5]
```diff
@@ -103,7 +113,15 @@ public void PropertyAccess(ILogger logger, Person person, List<Person> people)
...
-        logger.LogInformation("{PersonFirstName} {PersonAge}", someFirstName, person.Age);  // Compliant
+        logger.LogInformation("{PersonFirstName} {PersonAge}", someFirstName, person.Age);      // Compliant
+
+        var first = person.FirstName;
+        var last = person.LastName;
+        var name = person.LastName;
+        logger.LogInformation("Person: {FirstName} {LastName}", first, last);       // Compliant
+        logger.LogInformation("Person: {FirstName} {LastName}", last, first);       // FN
+        logger.LogInformation("Person: {FirstName} {LastName}", last, name);        // FN
+        logger.LogInformation("Person: {FirstName} {LastName}", name, name);        // FN
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/LoggingTemplatePlaceHoldersShouldBeInOrder.cs:124]: I think these are not really FNs. Just cases where the intention is not clear and we cannot tell.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs [hunk 6]
```diff
@@ -92,12 +94,52 @@ private static SyntaxNode OutOfOrderPlaceholderValue(MessageTemplatesParser.Plac
...
+            if (char.IsLetter(ch))
+            {
+                builder.Append(ch);
+            }
+        }
+        if (builder.Length > 0)
+        {
+            components.Add(builder.ToString());
+        }
+        return [.. components];
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs:140]: Isn't this copy the list contents to a new array? Why not return the list directly?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs [hunk 7]
```diff
@@ -92,12 +94,52 @@ private static SyntaxNode OutOfOrderPlaceholderValue(MessageTemplatesParser.Plac
...
+    {
+        if (isStrict)
+        {
+            return OnlyLetters(placeholderName).Equals(OnlyLetters(argumentName), StringComparison.OrdinalIgnoreCase);
+        }
+        else
+        {
+            var placeholderComponents = SplitByCamelCase(placeholderName);
+            var argumentComponents = SplitByCamelCase(argumentName);
+            return Array.Exists(placeholderComponents, x => Array.Exists(argumentComponents, y => x.Equals(y, StringComparison.OrdinalIgnoreCase)));
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs:116]: Isn't this the same with:
```suggestion
            return placeholderComponents.Intersect(argumentComponents, StringComparer.OrdinalIgnoreCase).Any();
```

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**